data/keystore/src/main/java/tech/pegasys/artemis/bls/keystore/KeyStore.java

```diff
@@ -0,0 +1,166 @@
+
+    final Crypto crypto = encryptUsingCipherFunction(blsPrivateKey, password, kdfParam, cipher);
+    final Bytes pubKey =
+        new PublicKey(SecretKey.fromBytes(Bytes48.leftPad(blsPrivateKey))).toBytesCompressed();
+    return new KeyStoreData(crypto, pubKey, path);
+  }
+
+  private static Crypto encryptUsingCipherFunction(
+      final Bytes secret, final String password, final KdfParam kdfParam, final Cipher cipher) {
+    final Bytes decryptionKey = kdfParam.generateDecryptionKey(password);
+    final Bytes cipherMessage =
+        applyCipherFunction(decryptionKey, cipher, true, secret.toArrayUnsafe());
+    final Bytes checksumMessage = calculateSHA256Checksum(decryptionKey, cipherMessage);
+    final Checksum checksum = new Checksum(checksumMessage);
+    final Cipher encryptedCipher =
+        new Cipher(cipher.getCipherFunction(), cipher.getCipherParam(), cipherMessage);
+    final Kdf kdf = new Kdf(kdfParam);
+    return new Crypto(kdf, checksum, encryptedCipher);
+  }
+
+  /**
+   * Validates password without decrypting the key as defined in specifications
+   *
+   * @param password The password to validate
+   * @param keyStoreData The Key Store against which password to validate
+   * @return true if password is valid, false otherwise.
+   */
+  public static boolean validatePassword(final String password, final KeyStoreData keyStoreData) {
+    checkNotNull(password, "Password cannot be null");
+    checkNotNull(keyStoreData, "KeyStoreData cannot be null");
+
+    final Bytes decryptionKey =
+        keyStoreData.getCrypto().getKdf().getParam().generateDecryptionKey(password);
+    return validateChecksum(decryptionKey, keyStoreData);
+  }
+
+  /**
+   * Decrypts BLS private key from the given KeyStore
+   *
+   * @param password The password to use for decryption
+   * @param keyStoreData The given Key Store
+   * @return decrypted BLS private key in Bytes
+   */
+  public static Bytes decrypt(final String password, final KeyStoreData keyStoreData) {
+    checkNotNull(password, "Password cannot be null");
+    checkNotNull(keyStoreData, "KeyStoreData cannot be null");
+
+    final Bytes decryptionKey =
+        keyStoreData.getCrypto().getKdf().getParam().generateDecryptionKey(password);
+
+    if (!validateChecksum(decryptionKey, keyStoreData)) {
+      throw new KeyStoreValidationException(
+          "Failed to decrypt KeyStore, checksum validation failed.");
+    }
+
+    final Cipher cipher = keyStoreData.getCrypto().getCipher();
+    final byte[] encryptedMessage = cipher.getMessage().toArrayUnsafe();
+    return applyCipherFunction(decryptionKey, cipher, false, encryptedMessage);
+  }
+
+  private static boolean validateChecksum(
+      final Bytes decryptionKey, final KeyStoreData keyStoreData) {
+    final Bytes checksum =
+        calculateSHA256Checksum(decryptionKey, keyStoreData.getCrypto().getCipher().getMessage());
+    return Objects.equals(checksum, keyStoreData.getCrypto().getChecksum().getMessage());
```

Comment:
Is there any reason you can't use the Bytes.equals method instead of Object.equals?

Reply:
They are essentially same, Objects.equals is slightly better approach as both parameters are checked instead of method chain like `checksum.equals(keystoredata...)`.